core/jobs/batch_jobs/mailchimp_population_jobs.py

```diff
@@ -0,0 +1,175 @@
+
+
+class CombineTuples(beam.CombineFn):  # type: ignore[misc]
+    """CombineFn for combining all user email, status tuples."""
+
+    def create_accumulator(self) -> List:
+        """Base accumulator where the tuples are added."""
+        return []
+
+    def add_input(self, accumulator: List, model: str) -> List:
+        """Append each tuple to the accumulator list."""
+        accumulator.append(model)
+        return accumulator
+
+    def merge_accumulators(self, accumulators: Iterable[List]) -> List:
+        """Merging accumulators is just combining both of them into a single
+        list.
+        """
+        return list(accumulators)
+
+    def extract_output(
+        self, accumulator: List
+    ) -> List:
+        """Output is the accumulator itself."""
+        return accumulator[0]
+
+
+class MailchimpPopulateJob(base_jobs.JobBase):
+    """One-off job for populating the mailchimp db."""
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        # Pcollection that returns the relevant config property with batch
+        # index.
+        config_property = (
+            self.pipeline
+            | 'Get all config properties' >> ndb_io.GetModels(
+                config_models.ConfigPropertyModel.get_all())
+            | 'Get the batch_index_for_mailchimp property value' >> beam.Filter(
+                lambda model: model.id == 'batch_index_for_mailchimp')
+            | 'Get value' >> beam.Map(lambda model: model.value)
+        )
+
+        batch_index_dict = beam.pvalue.AsSingleton(config_property)
+
+        # PCollection with all user ids that have opted in for email
+        # newsletters.
+        relevant_user_ids = (
+            self.pipeline
+            | 'Get all UserEmailPreferencesModel' >> ndb_io.GetModels(
+                user_models.UserEmailPreferencesModel.get_all().filter(
+                    user_models.UserEmailPreferencesModel.site_updates == True # pylint: disable=singleton-comparison
+                ).order('created_on'))
+            | 'Extract user id as tuple' >> beam.Map(
+                lambda preferences_model: preferences_model.id)
+        )
+
+        valid_user_ids = beam.pvalue.AsIter(relevant_user_ids)
+
+        # PCollection of all user emails opted in for newsletters.
+        relevant_user_emails = (
+            self.pipeline
+            | 'Get all user settings models' >> ndb_io.GetModels(
+                user_models.UserSettingsModel.get_all().order(
+                    'created_on'))
+            | 'Filter user models' >> (
+                beam.Filter(
+                    lambda model, ids: model.id in ids, ids=valid_user_ids))
+            | 'Get email' >> (beam.Map(lambda model: model.email))
+        )
+
+        mailchimp_results = (
+            relevant_user_emails
+            | 'Combine all valid emails into a single list' >>
+                beam.CombineGlobally(CombineTuples())
```

Comment:
I do not understand why do we need this combine now.

Reply:
The ParDo after this works on all emails at once to send a batch request. Without this combine, the DoFn will only get one email as its input parameter right?
Or is there a simpler way to have a function have all elements in the pcollection as its input.

Reply:
Done, used `beam.transforms.util.BatchElements` with a batch size of 50000 so that all emails are in a single list. The total user count on prod DB was only around 15k, so there is plenty of headroom here.